Fix ruby 2.2 comparable warnings

jgeiger opened the pull request:
Both changes fix issues with warning messages displayed.

The time_with_zone change doesn't appear to be a problem in master on Ruby 2.2, but it does cause the warning message issue in 3-2-stable.

seuros commented:
please squash commits.

senny commented:
Rails 3.2. is no longer maintained and we no longer merge patches for that branch.

@jgeiger Thank you for your work and your understanding.

jgeiger commented:
@senny Thanks for the info, but this is a replacement pull request for #18713 that was suggested by @rafaelfranca to remove warning messages caused by Ruby 2.2.

If you check that previous pull request, you can see that patches were/are being merged for Ruby 2.2 compatibility.

rafaelfranca commented:
:smile:

senny commented:
@jgeiger sorry, my bad :sweat:

rafaelfranca commented on activerecord/lib/active_record/fixtures.rb:
```diff
@@ -589,7 +589,7 @@ def table_rows
 
           # interpolate the fixture label
           row.each do |key, value|
-            row[key] = label if value == "$LABEL"
```
I believe we only need to do it if `value` is a string, so I think it is better to do:

```ruby
row[key] = label if value.is_a?(String) && value == "$LABEL"
```

rafaelfranca commented on activesupport/lib/active_support/time_with_zone.rb:
```diff
@@ -68,7 +68,7 @@ def period
     # Returns the simultaneous time in <tt>Time.zone</tt>, or the specified zone.
     def in_time_zone(new_zone = ::Time.zone)
-      return self if time_zone == new_zone
+      return self if new_zone == time_zone
```
The correct fix here is at `<=>` in `lib/active_support/values/time_zone.rb`.

We should return early if the argument doesn't respond to `utc_offset` and `name`.

rafaelfranca replied:
```diff
diff --git a/activesupport/lib/active_support/values/time_zone.rb b/activesupport/lib/active_support/values/time_zone.rb
index 01aea09..b5794be 100644
--- a/activesupport/lib/active_support/values/time_zone.rb
+++ b/activesupport/lib/active_support/values/time_zone.rb
@@ -223,6 +223,8 @@ module ActiveSupport
     # Compare this time zone to the parameter. The two are compared first on
     # their offsets, and then by name.
     def <=>(zone)
+      return unless zone.respond_to?(:utc_offset) && zone.respond_to?(:name)
+
       result = (utc_offset <=> zone.utc_offset)
       result = (name <=> zone.name) if result == 0
       result
```

Commit:
Fix ruby 2.2 comparable warnings

Check for correct value type in activerecord/fixtures.rb
Check that zone can respond to expected values to make the comparison.

jgeiger commented:
I've updated the changes as suggested and re pushed my branch.

Thanks.

rubydev commented:
I updated Rails and Ruby today to new versions: Rails 3.2.22 and Ruby to 2.2.0, but when I run tests I have many these warnings:

<pre>
/Users/marian/.rvm/gems/ruby-2.2.0@hospic/gems/activesupport-3.2.22/lib/active_support/callbacks.rb:408: warning: Comparable#== will no more rescue exceptions of #<=> in the next release.
/Users/marian/.rvm/gems/ruby-2.2.0@hospic/gems/activesupport-3.2.22/lib/active_support/callbacks.rb:408: warning: Return nil in #<=> if the comparison is inappropriate or avoid such comparison.
</pre>
